solana/solana-ibc/programs/solana-ibc/src/validation_context.rs

```diff
@@ -79,27 +81,40 @@ impl ibc::ValidationContext for IbcStorage<'_, '_> {
         height: &ibc::Height,
     ) -> Result<Self::AnyConsensusState> {
         let store = self.borrow();
-        let state = if height.revision_number() == 1 {
-            store.chain.consensus_state(height.revision_height().into())?
-        } else {
-            None
+        #[cfg(feature = "witness")]
+        {
+            let (_slot, fetched_timestamp, fetched_trie_root) = store
+                .private
+                .local_consensus_state
+                .iter()
+                .find(|cs| cs.0 == height.revision_height())
+                .unwrap();
+            let rollup_consensus_state = cf_solana::ConsensusState {
+                trie_root: ibc::CommitmentRoot::from_bytes(
+                    fetched_trie_root.as_slice(),
+                ),
+                timestamp_sec: NonZeroU64::new(*fetched_timestamp).unwrap(),
+            };
+            Ok(AnyConsensusState::Rollup(rollup_consensus_state))
+        }
+        #[cfg(not(feature = "witness"))]
+        {
+            let state = if height.revision_number() == 1 {
+                store.chain.consensus_state(height.revision_height().into())?
+            } else {
+                None
+            }
+            .ok_or(
+                ibc::ClientError::MissingLocalConsensusState {
+                    height: *height,
+                },
+            )?;
+            let guest_consensus_state = cf_guest::ConsensusState {
+                block_hash: state.0.as_array().to_vec().into(),
+                timestamp_ns: state.1,
+            };
+            return Ok(AnyConsensusState::Guest(guest_consensus_state));
         }
```

Comment:
Generation of the error could be move outside of the conditional compilation, though I’m not sure if this is better or worse. Up to you:

```suggestion
        #[cfg(feature = "witness")]
        let state = store
            .private
            .local_consensus_state
            .iter()
            .find(|cs| cs.0 == height.revision_height())
            .map(|(_slot, timestamp, trie_root)| {
                let state = cf_solana::ConsensusState {
                    trie_root: ibc::CommitmentRoot::from_bytes(
                        trie_root.as_slice(),
                    ),
                    timestamp_sec: NonZeroU64::new(*timestamp).unwrap(),
                };
                AnyConsensusState::Rollup(state)
            });
        #[cfg(not(feature = "witness"))]
        let state = (height.revision_number() == 1)
            .then(|| {
                let height = height.revision_height().into();
                store.chain.consensus_state(height)
            })
            .transpose()?
            .map(|state| {
                let state = cf_guest::ConsensusState {
                    block_hash: state.0.as_array().to_vec().into(),
                    timestamp_ns: state.1,
                };
                AnyConsensusState::Guest(state)
            });
        state.ok_or(ibc::ClientError::MissingLocalConsensusState {
            height: *height,
        })
```

Reply:
this looks clean.

Reply:
Also, CryptoHash has to_vec function so it might be possible to do:

    trie_root: trie_root.to_vec().into(),

which is more concise.

Reply:
got it